src/stan/model/indexing/assign_cl.hpp

```diff
@@ -0,0 +1,385 @@
+ * @param name Name of lvalue variable
+ * @param row_index index used for indexing rows of `expr_lhs`
+ * @param col_index index used for indexing columns of `expr_lhs`
+ * @throw std::out_of_range If the index is out of bounds.
+ */
+template <typename ExprLhs, typename ScalRhs,
+          require_rev_kernel_expression_t<ExprLhs>* = nullptr,
+          require_arithmetic_t<ScalRhs>* = nullptr>
+inline void assign(ExprLhs&& expr_lhs, const ScalRhs& scal_rhs,
+                   const char* name, const index_uni row_index,
+                   const index_uni col_index) {
+  decltype(auto) lhs = math::block_zero_based(
+      expr_lhs.val_op(), row_index.n_ - 1, col_index.n_ - 1, 1, 1);
+  math::arena_matrix_cl<double> prev_val = lhs;
+  lhs = math::constant(scal_rhs, 1, 1);  // assign the values
+  math::reverse_pass_callback(
+      [expr_lhs, row_index, col_index, prev_val]() mutable {
+        auto&& lhs_val = math::block_zero_based(
+            expr_lhs.val_op(), row_index.n_ - 1, col_index.n_ - 1, 1, 1);
+        auto&& lhs_adj = math::block_zero_based(
+            expr_lhs.adj(), row_index.n_ - 1, col_index.n_ - 1, 1, 1);
+        math::results(lhs_adj, lhs_val)
+            = math::expressions(math::constant(0.0, 1, 1), prev_val);
+      });
+}
+
+// assign(rev, rev, ...)
+/**
+ * Assign one reverse mode kernel generator expression to another, using given
+ * index.
+ *
+ * @tparam ExprLhs type of the expression on the left hand side of the
+ * assignment
+ * @tparam ExprRhs type of the expression on the right hand side of the
+ * assignment
+ * @tparam RowIndex type of index
+ * @param[in,out] expr_lhs expression on the left hand side of the assignment
+ * @param expr_rhs expression on the right hand side of the assignment
+ * @param name Name of lvalue variable
+ * @param row_index index used for indexing `expr_lhs`
+ * @throw std::out_of_range If the index is out of bounds.
+ * @throw std::invalid_argument If the right hand side size isn't the same as
+ * the indexed left hand side size.
+ */
+template <typename ExprLhs, typename ExprRhs, typename RowIndex,
+          require_all_rev_kernel_expression_t<ExprLhs, ExprRhs>* = nullptr>
+inline void assign(ExprLhs&& expr_lhs, const ExprRhs& expr_rhs,
+                   const char* name, const RowIndex& row_index) {
+  if (std::is_same<RowIndex, index_omni>::value) {
+    stan::math::check_size_match("omni assign", "left hand side rows",
+                                 expr_lhs.rows(), name, expr_rhs.rows());
+    stan::math::check_size_match("omni assign", "left hand side columns",
+                                 expr_lhs.cols(), name, expr_rhs.cols());
+  }
+  decltype(auto) lhs = rvalue(expr_lhs.val_op(), name, row_index);
+  math::arena_matrix_cl<double> prev_vals = lhs;
+  lhs = expr_rhs.val();  // assign the values
+  math::reverse_pass_callback(
+      [expr_lhs, expr_rhs, row_index, name, prev_vals]() mutable {
```

Comment:
Is `RowIndex` always trivially copyable? Might be good to include that in the docs if so

Reply:
No it is not. We need to forward it into lambda. Except that lambda captures do not support forwarding, so we need a workaround by wrapping it in a tuple ... Fixed.

Reply:
Sorry what's the logic with wrapping it in a tuple? I'm not sure I follow that

Reply:
So we can not use perfect forwarding, but we want to. So we pass in a tuple by value and forward the tuple element.

Reply:
Sorry which one is not trivially copyable? I'm looking at `index.hpp` and the only one I see is `index_multi` right? And the OpenCL assign / lvalues won't see that one because they are going to be pulled out as `matrix_cl<int>` types. So for that what if we just extended `to_arena()` for the indices that was essentially a no-op for the other indices and made a `matrix_cl<int>` in the arena when it sees `matrix_cl<int>`? The rest just have int values / default constructors / descructors, etc.

Reply:
Also why the need for perfect forwarding into the reverse pass callback? I mostly just care that we are sure stuff will exist and get properly cleaned up for the reverse pass

Reply:
Because `matrix_cl` is expensive to copy. Although what you suggest with to_arena should work as well and is simpler. Will change to that.